pilz_control/test/unittest_pilz_joint_trajectory_controller.cpp

```diff
@@ -112,191 +211,395 @@ void PilzJointTrajectoryControllerTest::SetUp()
-  progressInTime(period);
+  std_srvs::TriggerRequest req;
+  std_srvs::TriggerResponse resp;
+  EXPECT_TRUE(controller_->handleUnHoldRequest(req, resp));
+  EXPECT_FALSE(resp.success);
+
+  ros::Duration stop_duration{STOP_TRAJECTORY_DURATION_SEC + 2*DEFAULT_UPDATE_PERIOD_SEC};
+  progressInTime(stop_duration);
   updateController();
 
-  trajectory_action_client_.waitForResult();
-  EXPECT_EQ(control_msgs::FollowJointTrajectoryResult::INVALID_GOAL, trajectory_action_client_.getResult()->error_code)
-      << "Obtained error code " << trajectory_action_client_.getResult()->error_code << " for goal while in HOLD mode.";
+  EXPECT_TRUE(controller_->handleUnHoldRequest(req, resp));
+  EXPECT_TRUE(resp.success);
+  EXPECT_TRUE(checkForUnhold());
 }
 
+/**
+ * @tests{start_holding,
+ * Tests holding the controller.
+ * }
+ * @tests{no_execution_during_hold,
+ * Tests holding the controller.
+ * }
+ */
+TEST_F(PilzJointTrajectoryControllerTest, testHoldSuccessAfterUnhold)
+{
+  ASSERT_TRUE(performFullControllerStartup());
+
+  std_srvs::TriggerRequest req;
+  std_srvs::TriggerResponse resp;
+  // run async such that stop trajectory can be executed in the meantime
+  std::future<bool> hold_future = std::async(std::launch::async,
+                                             [this, &req, &resp]()
+                                             {
+                                               return controller_->handleHoldRequest(req, resp);
+                                             });
+
+  std::chrono::milliseconds hold_timeout {WAIT_FOR_HOLD_FUTURE_MSEC};
+  EXPECT_TRUE(updateControllerUntil([&hold_future](){ return isFutureReady(hold_future); }, hold_timeout));
+  EXPECT_TRUE(resp.success);
+  EXPECT_TRUE(checkForHold());
+}
 
 /**
- * @brief Test for correct return values and elapsed times in case of subsequent calls to hold/unhold function.
- *
- * Test Sequence:
- *    0. Start controller and perform initial update.
- *    1. Request DEFAULT twice.
- *    2. Request HOLDING.
- *    3. Request HOLDING again.
- *
- * Expected Results:
- *    0. Controller is executing (entering HOLD mode at startup).
- *    1. Each Response contains success==True.
- *    2. Response contains success==True. The elapsed time corresponds to the expected stop trajectory duration.
- *    3. Response contains success==True. The second call does not trigger a stop trajectory.
+ * @tests{start_holding,
+ * Tests holding the controller.
+ * }
+ * @tests{no_execution_during_hold,
+ * Tests holding the controller.
+ * }
  */
-TEST_F(PilzJointTrajectoryControllerTest, testDoubleRequest)
+TEST_F(PilzJointTrajectoryControllerTest, testDoubleHoldSuccess)
 {
   ASSERT_TRUE(controller_->init(hardware_, nh_, controller_nh_)) << "Failed to initialize the controller.";
+  ASSERT_TRUE(waitForActionServer());
   controller_->state_ = controller_->INITIALIZED;
-  startController();
 
+  startController();
   controller_->state_ = controller_->RUNNING;
```

Comment:
I thought, `unhold` is not necessary here and `performFullControllerStartup()` includes a `unhold`. Do you think we should still call it in order to avoid confusion?

Remember that the controller is in `stopping` mode after startup. When the first `hold` returns the controller has switched to `hold` mode.

Reply:
> Remember that the controller is in `stopping` mode after startup. When the first `hold` returns the controller has switched to `hold` mode. 

Ah, I forgot that we changed the initial mode to `stopping`. 

Anyway, what's the objective of the test? You wrote as test docu: "Tests triggering hold two times successively." If I read your docu I think that both `hold`'s are triggered while the "controller" is in idle state and not while the controller is still starting up. Therefore, I suggest to add the `performFullControllerStartup()` at the beginning.

Reply:
Ok.